File: android/core-export/src/main/kotlin/com/rfmapper/core/export/ExportEngineV1.kt
package com.rfmapper.core.export

import com.rfmapper.core.model.ExportKind
import com.rfmapper.core.model.ExportManifest
import com.rfmapper.core.model.Iso8601
import com.rfmapper.core.model.Observation
import com.rfmapper.core.model.ObserverIdentity
import com.rfmapper.core.model.RfMapperJson
import com.rfmapper.core.model.SampleKind
import com.rfmapper.core.model.SessionSummary
import com.rfmapper.core.model.csv.ObservationCsvCodec
import java.io.OutputStream

/**
 * Writes the V1 observation package described in `docs/16-export-package-specification.md`.
 *
 * Everything is streamed: each observation is encoded and written as it is pulled from the source,
 * so peak memory is independent of package size. The CSV and JSON entries each get their own pass
 * over the source (see [ObservationSource]), and the two passes are cross-checked so a disagreement
 * becomes an error rather than a package whose two views of the same data differ.
 */
class ExportEngineV1 : ExportEngine {

    override val version: String = "export_engine-1.0.0"

    override fun write(
        request: ExportRequest,
        source: ObservationSource,
        sink: ExportSink,
    ): ExportResult {
        val digests = LinkedHashMap<String, String>()
        val sizes = LinkedHashMap<String, Long>()

        val manifest = ExportPackage.manifestFor(request)
        writeEntry(sink, ExportPackage.MANIFEST, digests, sizes) { out ->
            out.writeUtf8(RfMapperJson.pretty.encodeToString(ExportManifest.serializer(), manifest))
            out.writeUtf8("\n")
        }

        val csvPass = StreamStatistics()
        writeEntry(sink, ExportPackage.OBSERVATIONS_CSV, digests, sizes) { out ->
            out.writeUtf8(ObservationCsvCodec.header)
            out.writeUtf8("\n")
            forEachOrdered(source) { observation ->
                csvPass.accept(observation)
                out.writeUtf8(ObservationCsvCodec.encode(observation))
                out.writeUtf8("\n")
            }
        }

        val jsonPass = StreamStatistics()
        writeEntry(sink, ExportPackage.OBSERVATIONS_JSON, digests, sizes) { out ->
            out.writeUtf8("[")
            forEachOrdered(source) { observation ->
                if (jsonPass.count > 0) out.writeUtf8(",")
                jsonPass.accept(observation)
                out.writeUtf8("\n")
                out.writeUtf8(RfMapperJson.compact.encodeToString(Observation.serializer(), observation))
            }
            if (jsonPass.count > 0) out.writeUtf8("\n")
            out.writeUtf8("]\n")
        }

        if (csvPass.rowFingerprint != jsonPass.rowFingerprint || csvPass.count != jsonPass.count) {
            throw ExportException(
                "the CSV and JSON passes disagree (${csvPass.count} vs ${jsonPass.count} rows): " +
                    "the observation source is not stable across passes",
            )
        }

        writeEntry(sink, ExportPackage.OBSERVER, digests, sizes) { out ->
            out.writeUtf8(RfMapperJson.pretty.encodeToString(ObserverIdentity.serializer(), request.observer))
            out.writeUtf8("\n")
        }

        if (request.sessions.isNotEmpty()) {
            writeEntry(sink, ExportPackage.SESSIONS, digests, sizes) { out ->
                val serializer = kotlinx.serialization.builtins.ListSerializer(SessionSummary.serializer())
                out.writeUtf8(RfMapperJson.pretty.encodeToString(serializer, request.sessions))
                out.writeUtf8("\n")
            }
        }

        verifyAgainstManifest(manifest, csvPass)

        // Written last, so its presence means every other entry completed.
        writeEntry(sink, ExportPackage.CHECKSUM, digests, sizes) { out ->
            out.writeUtf8(Sha256.checksumFile(digests))
        }

        sink.finish()

        return ExportResult(
            packageName = ExportPackage.fileName(
                observerId = request.observer.observerId,
                dateStamp = dateStampFor(request),
                sessionId = request.sessionIdForName.takeIf { request.exportKind == ExportKind.SESSION },
            ),
            observationsWritten = csvPass.count,
            entryDigests = digests,
            bytesPerEntry = sizes,
        )
    }

    private inline fun forEachOrdered(source: ObservationSource, action: (Observation) -> Unit) {
        val iterator = source.open()
        var previousTimestamp: String? = null
        var previousId: String? = null
        while (iterator.hasNext()) {
            val observation = iterator.next()
            if (previousTimestamp != null) {
                val comparison = compareValuesBy(
                    previousTimestamp to previousId!!,
                    observation.timestampUtc to observation.observationId,
                    { it.first },
                    { it.second },
                )
                if (comparison > 0) {
                    throw ExportException(
                        "observations must be ordered by (timestamp_utc, observation_id): " +
                            "$previousId preceded ${observation.observationId}",
                    )
                }
            }
            previousTimestamp = observation.timestampUtc
            previousId = observation.observationId
            action(observation)
        }
    }

    private fun dateStampFor(request: ExportRequest): String =
        request.dateRange?.from?.take(10)
            ?: request.summary.firstObservationUtc?.take(10)
            ?: Iso8601.utcDateStamp(request.createdAtEpochMillis)

    /**
     * Checks the precomputed manifest against what was actually streamed. A disagreement means the
     * database aggregates and the row query disagree, which would otherwise ship as a package whose
     * manifest quietly misstates its contents.
     */
    private fun verifyAgainstManifest(manifest: ExportManifest, observed: StreamStatistics) {
        if (manifest.observationCount != observed.count) {
            throw ExportException(
                "manifest declares ${manifest.observationCount} observations but ${observed.count} were written",
            )
        }
        if (observed.count > 0) {
            if (manifest.firstObservation != observed.firstTimestamp) {
                throw ExportException(
                    "manifest first_observation ${manifest.firstObservation} != streamed ${observed.firstTimestamp}",
                )
            }
            if (manifest.lastObservation != observed.lastTimestamp) {
                throw ExportException(
                    "manifest last_observation ${manifest.lastObservation} != streamed ${observed.lastTimestamp}",
                )
            }
        }
        if (manifest.countsBySensorType.isNotEmpty() &&
            manifest.countsBySensorType != observed.countsBySensorType
        ) {
            throw ExportException(
                "manifest counts_by_sensor_type ${manifest.countsBySensorType} != " +
                    "streamed ${observed.countsBySensorType}",
            )
        }
        if (manifest.groundTruthCount != observed.groundTruthCount) {
            throw ExportException(
                "manifest ground_truth_count ${manifest.groundTruthCount} != " +
                    "streamed ${observed.groundTruthCount}",
            )
        }
        observed.observerIds.firstOrNull { it != manifest.observerId }?.let { stray ->
            throw ExportException(
                "package declares observer ${manifest.observerId} but contains a row from $stray",
            )
        }
    }

    private inline fun writeEntry(
        sink: ExportSink,
        name: String,
        digests: MutableMap<String, String>,
        sizes: MutableMap<String, Long>,
        body: (OutputStream) -> Unit,
    ) {
        val raw = sink.openEntry(name)
        val digesting = DigestingOutputStream(raw)
        try {
            body(digesting)
            digesting.flush()
        } finally {
            raw.close()
        }
        digests[name] = digesting.digestHex()
        sizes[name] = digesting.bytesWritten
    }

    private fun OutputStream.writeUtf8(text: String) = write(text.toByteArray(Charsets.UTF_8))

    /**
     * Running statistics over a streamed pass, including a rolling fingerprint of the row sequence.
     * The fingerprint lets two passes be compared without retaining either of them.
     */
    private class StreamStatistics {
        var count: Long = 0
            private set
        var firstTimestamp: String? = null
            private set
        var lastTimestamp: String? = null
            private set
        var groundTruthCount: Long = 0
            private set
        var rowFingerprint: Long = FNV_OFFSET_BASIS
            private set

        val countsBySensorType = LinkedHashMap<String, Long>()
        val observerIds = LinkedHashSet<String>()

        fun accept(observation: Observation) {
            count++
            if (firstTimestamp == null) firstTimestamp = observation.timestampUtc
            lastTimestamp = observation.timestampUtc
            countsBySensorType.merge(observation.sensorType.name, 1L, Long::plus)
            if (observation.sampleKind == SampleKind.GROUND_TRUTH) groundTruthCount++
            observerIds += observation.observerId
            for (char in observation.observationId) {
                rowFingerprint = (rowFingerprint xor char.code.toLong()) * FNV_PRIME
            }
        }

        private companion object {
            const val FNV_OFFSET_BASIS = -3_750_763_034_362_895_579L
            const val FNV_PRIME = 1_099_511_628_211L
        }
    }
}

File: ios/Packages/RFMapper/Sources/RFMapperCore/ExportEngineV1.swift
/// Writes the V1 observation package described in `docs/16-export-package-specification.md`, byte
/// for byte as the Android Collector writes it.
///
/// One deliberate divergence from the Kotlin engine: that one streams, holding peak memory
/// independent of package size, and makes two passes over the row source specifically so neither
/// the CSV nor the JSON view has to be buffered. This one takes an array.
///
/// The reason is not laziness about memory but honesty about the platform. An iOS Collector cannot
/// run unattended for a day -- unfiltered BLE scanning is a foreground, screen-on activity, and
/// there is no equivalent of Android's foreground service (`docs/06-ios-capability-matrix.md` §4).
/// Its packages are supervised-session sized, thousands of rows rather than hundreds of thousands.
/// Streaming machinery sized for a volume the platform cannot produce would be speculative
/// complexity, and the two-pass cross-check it exists to enable is vacuous when the rows are already
/// a value that cannot change between passes.
public struct ExportEngineV1 {

    public static let version = "export_engine-1.0.0"

    public init() {}

    public func write(request: ExportRequest, observations: [Observation]) throws -> ExportResult {
        try request.summary.validate()
        try request.observer.validate()

        let manifest = ExportPackage.manifest(for: request)
        try manifest.validate()

        try verifyOrdering(observations)

        var writer = ZipWriter()
        var digests: [String: String] = [:]
        var sizes: [String: Int] = [:]

        func writeEntry(_ name: String, _ text: String) {
            let bytes = Array(text.utf8)
            writer.addEntry(name: name, bytes: bytes)
            digests[name] = Sha256.hex(bytes)
            sizes[name] = bytes.count
        }

        writeEntry(ExportPackage.manifest, CanonicalJSON.pretty(manifest.json) + "\n")

        var csv = ObservationCsvCodec.header
        csv += "\n"
        for observation in observations {
            csv += ObservationCsvCodec.encode(observation)
            csv += "\n"
        }
        writeEntry(ExportPackage.observationsCsv, csv)

        // The array layout the Kotlin engine produces: an opening bracket, then each row on its own
        // line as compact JSON, separated by commas. Not a pretty-printed array -- the rows are
        // compact so that a package stays greppable one observation at a time.
        var json = "["
        for (index, observation) in observations.enumerated() {
            if index > 0 { json += "," }
            json += "\n"
            json += observation.compactJSON
        }
        if !observations.isEmpty { json += "\n" }
        json += "]\n"
        writeEntry(ExportPackage.observationsJson, json)

        writeEntry(ExportPackage.observer, CanonicalJSON.pretty(request.observer.json) + "\n")

        if !request.sessions.isEmpty {
            let array = JSONValue.array(request.sessions.map(\.json))
            writeEntry(ExportPackage.sessions, CanonicalJSON.pretty(array) + "\n")
        }

        let statistics = Statistics(observations)
        try verify(manifest: manifest, against: statistics)

        // Written last, so its presence means every other entry completed.
        writeEntry(ExportPackage.checksum, Sha256.checksumFile(digests))

        return ExportResult(
            packageName: ExportPackage.fileName(
                observerId: request.observer.observerId,
                dateStamp: dateStamp(for: request),
                sessionId: request.exportKind == .session ? request.sessionIdForName : nil
            ),
            bytes: writer.finish(),
            observationsWritten: statistics.count,
            entryDigests: digests,
            bytesPerEntry: sizes
        )
    }

    private func verifyOrdering(_ observations: [Observation]) throws {
        var previous: (timestamp: String, id: String)?
        for observation in observations {
            if let previous {
                let ordered = (previous.timestamp, previous.id)
                    <= (observation.timestampUtc, observation.observationId)
                if !ordered {
                    throw ExportError.unorderedRows(
                        previous: previous.id,
                        next: observation.observationId
                    )
                }
            }
            previous = (observation.timestampUtc, observation.observationId)
        }
    }

    private func dateStamp(for request: ExportRequest) -> String {
        if let from = request.dateRange?.from {
            return String(from.prefix(10))
        }
        if let first = request.summary.firstObservationUtc {
            return String(first.prefix(10))
        }
        return Iso8601.utcDateStamp(request.createdAtEpochMillis)
    }

    /// Checks the precomputed manifest against what was actually written. A disagreement means the
    /// aggregates and the rows disagree, which would otherwise ship as a package whose manifest
    /// quietly misstates its contents.
    private func verify(manifest: ExportManifest, against observed: Statistics) throws {
        if manifest.observationCount != observed.count {
            throw ExportError.manifestDisagreement(
                "manifest declares \(manifest.observationCount) observations but "
                    + "\(observed.count) were written"
            )
        }
        if observed.count > 0 {
            if manifest.firstObservation != observed.firstTimestamp {
                throw ExportError.manifestDisagreement(
                    "manifest first_observation \(manifest.firstObservation ?? "nil") != "
                        + "written \(observed.firstTimestamp ?? "nil")"
                )
            }
            if manifest.lastObservation != observed.lastTimestamp {
                throw ExportError.manifestDisagreement(
                    "manifest last_observation \(manifest.lastObservation ?? "nil") != "
                        + "written \(observed.lastTimestamp ?? "nil")"
                )
            }
        }
        if !manifest.countsBySensorType.isEmpty,
           manifest.countsBySensorType != observed.countsBySensorType {
            throw ExportError.manifestDisagreement(
                "manifest counts_by_sensor_type \(manifest.countsBySensorType) != "
                    + "written \(observed.countsBySensorType)"
            )
        }
        if manifest.groundTruthCount != observed.groundTruthCount {
            throw ExportError.manifestDisagreement(
                "manifest ground_truth_count \(manifest.groundTruthCount) != "
                    + "written \(observed.groundTruthCount)"
            )
        }
        if let stray = observed.observerIds.sorted().first(where: { $0 != manifest.observerId }) {
            throw ExportError.manifestDisagreement(
                "package declares observer \(manifest.observerId) but contains a row from \(stray)"
            )
        }
    }

    private struct Statistics {
        var count: Int64 = 0
        var firstTimestamp: String?
        var lastTimestamp: String?
        var groundTruthCount: Int64 = 0
        var countsBySensorType: [String: Int64] = [:]
        var observerIds: Set<String> = []

        init(_ observations: [Observation]) {
            for observation in observations {
                count += 1
                if firstTimestamp == nil { firstTimestamp = observation.timestampUtc }
                lastTimestamp = observation.timestampUtc
                countsBySensorType[observation.sensorType.rawValue, default: 0] += 1
                if observation.sampleKind == .groundTruth { groundTruthCount += 1 }
                observerIds.insert(observation.observerId)
            }
        }
    }
}
